02-sat-advanced/homework-students-solutions/elblasco_hw_221.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+
+# https://logic.puzzlebaron.com/play.php?u2=9db6a5914e4ef572fe50bc57bac9efd5
+
+from pysmt.shortcuts import (
+    Not, Or, And, ExactlyOne, Solver, Symbol, BOOL, TRUE
+)
+
+prices = ["45", "60", "75", "90", "105"]
+winners = ["alejandro", "brandon", "linda", "nick", "walter"]
+butterflies = ["atlas", "clearwing", "grayling", "peacock", "peppered"]
+
+wvars = {
+    f"{price}{winner}": Symbol(f"{price}{winner}", BOOL)
+    for price in prices
+    for winner in winners
+}
+
+bvars = {
+    f"{price}{butterfly}": Symbol(f"{price}{butterfly}", BOOL)
+    for price in prices
+    for butterfly in butterflies
+}
+
+assertions = []
+
+# 1. Walter's purchase didn't cost $105.
+assertions.append(
+    Not(wvars["105walter"])
+)
+
+# 2. The five butterflies were Nick's purchase, the butterfly that sold for $45
+# the peacock butterfly, the butterfly that sold for $90 and the butterfly that
+# sold for $105.
+
+# Nick did not do any purchase of 45$, 90$ or 105$
+assertions.append(
+    Not(
+        Or(
+            wvars["45nick"], wvars["90nick"], wvars["105nick"]
+        )
+    )
+)
+
+# The peacock has not been bought for 45$, 90$ or 105$
+assertions.append(
+    Not(
+        Or(
+            bvars["45peacock"], bvars["90peacock"], bvars["105peacock"]
+        )
+    )
+)
+
+# 3. The grayling butterfly cost 15 dollars less than the atlas butterfly.
+assertions.append(
+    Or(
+        And(bvars["45grayling"], bvars["60atlas"]),
+        And(bvars["60grayling"], bvars["75atlas"]),
+        And(bvars["75grayling"], bvars["90atlas"]),
+        And(bvars["90grayling"], bvars["105atlas"]),
+    )
+)
+
+# 4. Linda's purchase was either the insect that sold for $60 or the grayling
+# butterfly.
+
+# Formula to encode "Linda bought grayling for a price which was not 60$"
+linda_bought_grayling = Or(
+    And(
+        wvars[f"{price}linda"], bvars[f"{price}grayling"]
+    )
+    for price in prices
+    if price != "60"
+)
+
+assertions.append(
+    Or(
+        And(wvars["60linda"], Not(linda_bought_grayling)),
```

Comment:
probably you can just say
```suggestion
        And(wvars["60linda"], Not(bvars["60grayling"])),
```